neuralmonkey/trainers/cross_entropy_trainer.py

```diff
@@ -21,10 +24,15 @@ def xent_objective(decoder, weight=None) -> Objective:
 
 class CrossEntropyTrainer(GenericTrainer):
 
-    def __init__(self, decoders: List[Any],
+    def __init__(self,
+                 decoders: List[Any],
                  decoder_weights: List[ObjectiveWeight] = None,
-                 l1_weight=0., l2_weight=0.,
-                 clip_norm=False, optimizer=None, global_step=None,
+                 l1_weight: float = 0.,
+                 l2_weight: float = 0.,
+                 clip_norm: float = None,
+                 optimizer_getter: OptimizerGetter = adam_optimizer(),
+                 global_step: tf.Tensor = None,
```

Comment:
Já bych asi ten `global_step` všude nepředával a použil [`tf.train.get_or_create_global_step()`](https://www.tensorflow.org/api_docs/python/tf/train/get_or_create_global_step). To se v případě potřeby dá zavolat i z konfiguráku, např.:

```ini
[lr_decay]
class=tf.train.exponential_decay
learning_rate=0.1
decay_steps=100000
decay_rate=0.96
global_step=<global_step>

[global_step]
class=tf.train.get_or_create_global_step
```

A námi definované decay funkce ten `global_step` mít jako povinný parametr nemusejí a zjistí si ho samy.

Reply:
Supr, tohle jsem neznal :)